src/ImageSharp/Formats/PixelTypeInfo.cs

```diff
@@ -4,50 +4,51 @@
 using System.Runtime.CompilerServices;
 using SixLabors.ImageSharp.PixelFormats;
 
-// TODO: Review this class as it's used to represent 2 different things.
+// TODO: Review this type as it's used to represent 2 different things.
 // 1.The encoded image pixel format.
 // 2. The pixel format of the decoded image.
 namespace SixLabors.ImageSharp.Formats;
 
 /// <summary>
 /// Contains information about the pixels that make up an images visual data.
 /// </summary>
-public class PixelTypeInfo
+/// <remarks>
+/// Initializes a new instance of the <see cref="PixelTypeInfo"/> struct.
+/// </remarks>
+/// <param name="bitsPerPixel">Color depth, in number of bits per pixel.</param>
+public readonly struct PixelTypeInfo(int bitsPerPixel)
 {
     /// <summary>
-    /// Initializes a new instance of the <see cref="PixelTypeInfo"/> class.
+    /// Gets color depth, in number of bits per pixel.
     /// </summary>
-    /// <param name="bitsPerPixel">Color depth, in number of bits per pixel.</param>
-    public PixelTypeInfo(int bitsPerPixel)
-        => this.BitsPerPixel = bitsPerPixel;
+    public int BitsPerPixel { get; init; } = bitsPerPixel;
 
     /// <summary>
-    /// Initializes a new instance of the <see cref="PixelTypeInfo"/> class.
+    /// Gets the count of the color components
     /// </summary>
-    /// <param name="bitsPerPixel">Color depth, in number of bits per pixel.</param>
-    /// <param name="alpha">The pixel alpha transparency behavior.</param>
-    public PixelTypeInfo(int bitsPerPixel, PixelAlphaRepresentation alpha)
-    {
-        this.BitsPerPixel = bitsPerPixel;
-        this.AlphaRepresentation = alpha;
-    }
+    public byte ComponentCount { get; init; }
 
     /// <summary>
-    /// Gets color depth, in number of bits per pixel.
+    /// Gets the pixel component precision.
     /// </summary>
-    public int BitsPerPixel { get; }
+    public PixelComponentPrecision? ComponentPrecision { get; init; }
```

Comment:
AFAIK there might be pixel types with assymmetric component precision. (I remember briefly touching this topic in an old discussion with Clinton.)

How would this model handle such pixel types?

Reply:
I looked at the usages of the property, and the only way we use it is the `info.ComponentPrecision <= PixelComponentPrecision.Byte` comparison, meaning that the only thing we need this for is to determine if the precision is below or above a certain treshold, and the exact precision information is practically unused. Unless we think there are use cases (proven by actual user user need) to observe *the precision of the highest precision component*, IMO we should try to avoid exposing such a property and explore alternative options to solve this.

Reply:
My thought process is that users can implement custom performance optimized processors based upon this implementation. If they know the precision, they could operate over `byte` vs `Vector4`. 

The use case that I've currently implemented fixes a bug we had also.

`Rgba1010102` uses a `uint` as its packed value which is the same size as `Rgba32` but we lose precision for the RGB components if we convert to it as we were since they require 10 bit precision.

Reply:
I see that `ComponentCount` is a similar informative property. I didn't notice this fact in the #2534 discussion. The way I view this now is that we are attempting to add a *Rich Pixel Type Descriptor* feature. IMO there are a couple of important points we should consider before fully committing to it:
- To really get this right, it might be better to expose a `struct/class ComponentInfo` and a property in `PixelTypeInfo` that enumerates the ComponentInfos, so each component can tell its precision separately instead of telling the maximum (for which I'm not sure if there is a use-case except the conversion implementation this PR is proposing). As far as I can recall that old discussion with Clinton, WIC has such a model but I may be wrong.
- Note that this feature idea doesn't necessarily overlap with the `Pixel <=> Color` conversion. I don't see how can we implement efficient bulk conversion using these descriptor information only. In fact, if our goal is to solve the `Pixel <=> Color` conversion thing, it might be better to promote the conversion (& bulk conversion) methods to `PixelOperations<T>`.
- Considering that this isn't super trivial to get right, I'm really curious if we have enough evidence (=enough users) proving that the  *Rich Pixel Type Descriptor* feature is worth the headache.

Reply:
(Your comment somehow didn't appear before me posting the previous one.)

> My thought process is that users can implement custom performance optimized processors based upon this implementation. 

Do you have any specific examples?

Reply:
>Considering that this isn't super trivial to get right, I'm really curious if we have enough evidence (=enough users) proving that the Rich Pixel Type Descriptor feature is worth the headache.

We get zero feedback on this stuff. I'm just trying to create something that gives us as much knowledge as possible. 

For example, We, could be using both the `PixelComponentPrecision` and `PixelAlphaRepresentation` properties in encoders to determine color types and bit depth no explicit encoding options are passed. This helps solve handling of introduced transparency in a processing operation.

Reply:
Here's the info from WIC

https://learn.microsoft.com/en-us/windows/win32/wic/-wic-codec-native-pixel-formats

And here's how it's dealt with in MagicScaler

https://github.com/saucecontrol/PhotoSauce/blob/v0.14.0/src/MagicScaler/Core/PixelFormats.cs

We're missing a color type enum which would give us a fairly robust model that can allow encoders to make sensible decisions.

Reply:
Here's an idea for how we can handle the pixel color type. We can combine them in any order to show component order but also test against them to determine encoding color type. 

```c#
[Flags]
public enum PixelColorType
{
  Red = 1,
  Green = 2,
  Blue = 4,
  Alpha = 8,
  Grayscale = 16
}
```

Defining them.

``` c#
var rgb = PixelColorType.Red | PixelColorType.Green | PixelColorType.Blue;
var  bgr = PixelColorType.Blue | PixelColorType.Green | PixelColorType.Red;
var la = PixelColorType.Grayscale | PixelColorType.Alpha;
```

Testing them.

```c#
info.ColorType.HasFlag(PixelColorType.Blue | PixelColorType.Green | PixelColorType.Red | PixelColorType.Alpha);
```

Reply:
> * To really get this right, it might be better to expose a `struct/class ComponentInfo` and a property in `PixelTypeInfo` that enumerates the ComponentInfos, so each component can tell its precision separately instead of telling the maximum (for which I'm not sure if there is a use-case except the conversion implementation this PR is proposing). As far as I can recall that old discussion with Clinton, WIC has such a model but I may be wrong.

You are quite correct.  I don't currently have this in my implementation because I haven't built my own converters for packed formats yet, but WIC does define per-channel masks so that you can build generic converters.  https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nf-wincodec-iwicpixelformatinfo-getchannelmask

Reply:
Whoa! How does that work?

Reply:
It's no different than what you've already got for decoding/converting BMP (https://github.com/SixLabors/ImageSharp/blob/main/src/ImageSharp/Formats/Bmp/BmpDecoderCore.cs#L1129) since it also allows arbitrary channel depths and masks.

WIC allows you to define/register your own optimized converters for formats you register, but otherwise it falls back to a generic one that just works off the defined masks.

Reply:
> It's no different than what you've already got for decoding/converting BMP

Except that the channel count is not limited to 4!

> so that you can build generic converters

Our "almost generic (tm)" way of doing this is to push everything through `Vector4`. Defining a pixel type in ImageSharp means implementing conversion to/from Vector4 and virtually all of our processors depend on Vector4-convertibility. We should decide what to do with the cases that don't fit this model. Maybe there are some limited applications for them, even though such images are for example "not resizable".

@saucecontrol I wonder if there are any image formats that can encode `GUID_WICPixelFormat144bpp8ChannelsAlpha`? What are real-life applications of such wild stuff? :)

Reply:
In general, it's a pretty safe assumption anything weird defined by WIC is to support the JXR/HD Photo codec https://learn.microsoft.com/en-us/windows/win32/wic/-wic-codec-native-pixel-formats#jpeg-xr-native-codec 😄 

I think Microsoft meant JXR to be all things to all people and implemented a lot of things that never actually got used.  In the case of the 6+ channel formats, I assume those were meant to replace some specialized uses of TIFF, since TIFF also allows an arbitrary number of channels.  The only real world use I can think of for those formats is print workflows since printers may have any number of inks.

WIC does support up to 5 channels in its TIFF codec, for CMYKA.  That's a real-world format, although even Photoshop will warn you that most software doesn't support it when you save in that format.  Here's a sample for reference:

[cmyka.zip](https://github.com/SixLabors/ImageSharp/files/13665156/cmyka.zip)

Windows photo viewer will open that one and display the transparency correctly, but its color management doesn't support CMYKA, so the colors will display incorrectly.  Using the Little-CMS plugin, I'm able to convert those correctly in MagicScaler, but other than Photoshop, I don't know offhand what else works.  Probably ImageMagick and GIMP?

Reply:
How about this? It stores up to 8 components with a precision of 255 bits per component. 

Usage 

<img  alt="image" src="https://github.com/SixLabors/ImageSharp/assets/385879/bc4fb0b4-c0dc-41ce-aec0-5dc8291764b7">


```c#
public readonly struct PixelPrecision
{
    private readonly int count;
    private readonly long precisionData;

    private PixelPrecision(int count, long precisionData)
    {
        this.count = count;
        this.precisionData = precisionData;
    }

    public int ComponentCount => this.count;

    public static PixelPrecision Create(int count, params int[] precision)
    {
        if (precision.Length != count || precision.Length > 8)
        {
            throw new ArgumentException($"Cannot have more than 8 components and must match {count}.");
        }

        long precisionData = 0;
        for (int i = 0; i < precision.Length; i++)
        {
            if (precision[i] is < 0 or > 255)
            {
                throw new ArgumentException("Precision must be between 0 and 255.");
            }

            precisionData |= ((long)precision[i]) << (8 * i);
        }

        return new PixelPrecision(count, precisionData);
    }

    public int GetComponentPrecision(int componentIndex)
    {
        if (componentIndex < 0 || componentIndex >= this.count)
        {
            throw new ArgumentOutOfRangeException($"index must be greater than or equal to 0 and less than {this.count}.");

        }

        return (int)((this.precisionData >> (8 * componentIndex)) & 0xFF);
    }

    public int GetMaximumComponentPrecision()
    {
        int maxPrecision = 0;
        for (int i = 0; i < this.count; i++)
        {
            int componentPrecision = (int)((this.precisionData >> (8 * i)) & 0xFF);
            if (componentPrecision > maxPrecision)
            {
                maxPrecision = componentPrecision;
            }
        }

        return maxPrecision;
    }
}
```

Reply:
Damn. Just realized `GUID_WICPixelFormat144bpp8ChannelsAlpha` has 9 channels!!! I guess I could use 2 backing stores?

Reply:
Here is as a class supporting up to 16 components. Each component precision maxes out at 255bits. 

I'm allowing such high precision because of https://github.com/SixLabors/ImageSharp.Drawing/discussions/165 
otherwise, I'd limit it to 16 bits and using a single backing store.

```c#
public sealed class PixelPrecision
{
    private readonly long precisionData1;
    private readonly long precisionData2;

    private PixelPrecision(int count, long precisionData1, long precisionData2)
    {
        this.ComponentCount = count;
        this.precisionData1 = precisionData1;
        this.precisionData2 = precisionData2;
    }

    public int ComponentCount { get; }

    public static PixelPrecision Create(int count, params int[] precision)
    {
        if (precision.Length != count || precision.Length > 16)
        {
            throw new ArgumentException($"Count must match the length of precision array and cannot exceed 16.");
        }

        long precisionData1 = 0;
        long precisionData2 = 0;
        for (int i = 0; i < precision.Length; i++)
        {
            if (precision[i] is < 0 or > 255)
            {
                throw new ArgumentException("Precision must be between 0 and 255.");
            }

            if (i < 8)
            {
                precisionData1 |= ((long)precision[i]) << (8 * i);
            }
            else
            {
                precisionData2 |= ((long)precision[i]) << (8 * (i - 8));
            }
        }

        return new PixelPrecision(count, precisionData1, precisionData2);
    }

    public int GetComponentPrecision(int componentIndex)
    {
        if (componentIndex < 0 || componentIndex >= this.ComponentCount)
        {
            throw new ArgumentOutOfRangeException($"Component index must be between 0 and {this.ComponentCount - 1} inclusive.");
        }

        long selectedPrecisionData = componentIndex < 8 ? this.precisionData1 : this.precisionData2;
        return (int)((selectedPrecisionData >> (8 * (componentIndex & 7))) & 0xFF);
    }

    public int GetMaximumComponentPrecision()
    {
        int maxPrecision = 0;
        for (int i = 0; i < this.ComponentCount; i++)
        {
            int componentPrecision = GetComponentPrecision(i);
            if (componentPrecision > maxPrecision)
            {
                maxPrecision = componentPrecision;
            }
        }

        return maxPrecision;
    }
}
```

Reply:
That's neat!  I think you might have to be able to account for padding bits, so I'm not sure you can get away with not having a mask per channel.  I'm trying to think of tricky examples... I've seen some BMPs that have gaps in the channel masks, but those were of the stress-test variety, not real world.  You'll have real-world formats like B10G10R10, which is 32-bit with 2 bits padding, though.

Reply:
I'm not sure that's required for our use cases which is what precision can I safly store this pixel in. 

We can tell the actual size of the pixel format from the info since the types themselves must handle padding. `Unsafe.Sizeof<B10G10R10>()` should be the same size `Unsafe.Sizeof<Rgba1010102>()` which is `4`.

Theoretically we could add padding to the info. I'm assuming though that world is not completely mad and padding is always at the end?

Reply:
Right, I don't think you'll see a real-world case where `BitsPerPixel - Sum(ChannelPrecision)` doesn't give you the padding bits, which *should* always be at the end.  Just trying to come up with reasons explicit masks might be beneficial, and weirdo pixel formats with interior padding are at least a possibility.

The other thing I guess that's useful in an explicit channel mask is that it also communicates the channel order.  I'm not sure if WIC actually uses that information in that way, but I suppose it might.  They don't expose channel order in any other way in their format definitions.  It's just in the name and in the mask.

Reply:
I think I'll go with real world for now. If someone wants something really weird, they can find a really weird library.

Reply:
One thing that does concern me though is the the performance of the parent type. We removed all the lazy loading as the type was to be a small and fast struct but now, I'm concerned about overheads.

Reply:
Type has been introduced as `PixelComponentInfo`